packages/grafana-ui/src/components/Table/Table.tsx

```diff
@@ -393,4 +432,115 @@ export const Table: FC<Props> = memo((props: Props) => {
   );
 });
 
+function useMainData(
+  data: DataFrame,
+  // This is index after filtering so we cannot use it directly with data
+  expandedIndex: number | undefined
+): { mainData: DataFrame; subData?: DataFrame; indexesToKeepBlank: number[]; expandable: boolean } {
+  return useMemo(() => {
+    const subCols: Field[] = [];
+    const mainCols: Field[] = [];
+    let indexesToKeepBlank: number[] = [];
+
+    for (const f of data.fields) {
+      if (f.config.custom?.subcol) {
+        subCols.push(f);
+      } else {
+        mainCols.push(f);
+      }
+    }
+
+    if (!subCols.length) {
+      return {
+        mainData: data,
+        subData: undefined,
+        indexesToKeepBlank,
+        expandable: false,
+      };
+    }
+
+    // We need to filter out rows and there does not seem to be an easy util for that. This is a bit convoluted though.
+    // Probably try using DataFrameView and MutableDataFrame to loop over rows instead of this manual indexing.
+
+    let inSubColumns = false;
+    const newMainValues = mainCols.map(() => new ArrayVector());
+    const newSubValues = subCols.map(() => new ArrayVector());
+
+    for (let rowIndex = 0; rowIndex < mainCols[0].values.length; rowIndex++) {
+      const isMainRow = mainCols.some((col) => col.values.get(rowIndex));
+      if (isMainRow) {
+        for (let columnIndex = 0; columnIndex < mainCols.length; columnIndex++) {
+          newMainValues[columnIndex].add(mainCols[columnIndex].values.get(rowIndex));
+        }
+
+        inSubColumns = false;
+        continue;
+      }
+
+      const isFirstExpandedSubRow =
+        !isMainRow && expandedIndex !== undefined && newMainValues[0].length - 1 === expandedIndex;
+      if (isFirstExpandedSubRow || inSubColumns) {
```

Comment:
Because we are using a virtual scroll view which needs items with the same height we cannot just add a row which is some random height. So we count how many rows will be in the sub table, add some buffer and return a data frame with empty rows to account for the space where the sub table will be rendered.